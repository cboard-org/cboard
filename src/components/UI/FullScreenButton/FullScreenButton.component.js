import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import FullscreenIcon from '@material-ui/icons/Fullscreen';
import FullscreenExitIcon from '@material-ui/icons/FullscreenExit';

import IconButton from '../IconButton';
import messages from './FullScreenButton.messages';

const propTypes = {
  /**
   * If true, button will be disabled
   */
  disabled: PropTypes.bool,
  /**
   * @ignore
   */
  intl: intlShape.isRequired
};

class FullScreenButton extends PureComponent {
  state = { fullscreen: false };

  requestFullscreen(element) {
    if (element.requestFullscreen) {
      element.requestFullscreen();
    } else if (element.mozRequestFullScreen) {
      element.mozRequestFullScreen();
    } else if (element.webkitRequestFullScreen) {
      element.webkitRequestFullscreen(Element.ALLOW_KEYBOARD_INPUT);
    }
    this.setState({ fullscreen: true });
  }

  exitFullscreen() {
    if (document.exitFullscreen) {
      document.exitFullscreen();
    } else if (document.mozCancelFullScreen) {
      document.mozCancelFullScreen();
    } else if (document.webkitExitFullscreen) {
      document.webkitExitFullscreen();
    }
    this.setState({ fullscreen: false });
  }

  toggleFullscreen() {
    if (this.state.fullscreen) {
      this.exitFullscreen();
    } else {
      this.requestFullscreen(window.document.documentElement);
    }
  }

  handleClick = () => {
    this.toggleFullscreen();
  };

  render() {
    const { disabled, intl } = this.props;

    const fullScreenLabel = this.state.fullscreen
      ? intl.formatMessage(messages.exitFullscreen)
      : intl.formatMessage(messages.fullscreen);

    return (
      <IconButton
        disabled={disabled}
        label={fullScreenLabel}
        onClick={this.handleClick}
      >
        {this.state.fullscreen ? <FullscreenExitIcon /> : <FullscreenIcon />}
      </IconButton>
    );
  }
}

FullScreenButton.propTypes = propTypes;

export default injectIntl(FullScreenButton);
