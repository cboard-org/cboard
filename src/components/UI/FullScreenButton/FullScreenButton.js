import React, { useState, useCallback, useEffect } from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import FullscreenIcon from '@material-ui/icons/Fullscreen';
import FullscreenExitIcon from '@material-ui/icons/FullscreenExit';

import IconButton from '../IconButton';
import messages from './FullScreenButton.messages';

const FullScreenButton = ({ disabled, intl }) => {
  const [fullscreen, setFullscreen] = useState(false);
  const requestFullScreen = useCallback(element => {
    let request =
      element.requestFullscreen ||
      element.webkitRequestFullscreen ||
      element.mozRequestFullScreen;
    if (request) {
      request.call(element);
      setFullscreen(true);
    }
  }, []);

  const exitFullscreen = useCallback(() => {
    let request =
      document.exitFullscreen ||
      document.webkitExitFullscreen ||
      document.mozCancelFullScreen;
    if (request) {
      request.call(document);
      setFullscreen(false);
    }
  }, []);
  const toggleFullScreen = useCallback(
    () => {
      if (fullscreen) {
        exitFullscreen();
      } else {
        requestFullScreen(document.documentElement);
      }
    },
    [fullscreen, requestFullScreen, exitFullscreen]
  );

  const handleClick = useCallback(
    () => {
      toggleFullScreen();
    },
    [toggleFullScreen]
  );

  useEffect(() => {
    const handleFullscreenChange = () => {
      setFullscreen(
        document.fullscreenElement ||
          document.webkitFullscreenElement ||
          document.mozFullScreenElement
      );
    };

    document.addEventListener('fullscreenchange', handleFullscreenChange);
    document.addEventListener('webkitfullscreenchange', handleFullscreenChange);
    document.addEventListener('mozfullscreenchange', handleFullscreenChange);

    return () => {
      document.removeEventListener('fullscreenchange', handleFullscreenChange);
      document.removeEventListener(
        'webkitfullscreenchange',
        handleFullscreenChange
      );
      document.removeEventListener(
        'mozfullscreenchange',
        handleFullscreenChange
      );
    };
  }, []);

  const fullScreenLabel = fullscreen
    ? intl.formatMessage(messages.exitFullscreen)
    : intl.formatMessage(messages.fullscreen);

  return (
    <IconButton
      disabled={disabled}
      label={fullScreenLabel}
      onClick={handleClick}
    >
      {fullscreen ? <FullscreenExitIcon /> : <FullscreenIcon />}
    </IconButton>
  );
};

FullScreenButton.propTypes = {
  disabled: PropTypes.bool,

  intl: intlShape.isRequired
};

export default injectIntl(FullScreenButton);
