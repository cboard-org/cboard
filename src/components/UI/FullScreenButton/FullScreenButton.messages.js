import { defineMessages } from 'react-intl';

export default defineMessages({
  fullscreen: {
    id: 'cboard.components.FullScreenButton.fullscreen',
    defaultMessage: 'Full screen'
  },
  exitFullscreen: {
    id: 'cboard.components.FullScreenButton.exitFullscreen',
    defaultMessage: 'Exit full screen'
  }
});
