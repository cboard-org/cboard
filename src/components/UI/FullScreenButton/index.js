export { default } from './FullScreenButton.component';
