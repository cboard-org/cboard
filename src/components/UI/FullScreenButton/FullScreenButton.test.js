import React from 'react';
import { shallowMatchSnapshot } from '../../../common/test_utils';

import FullScreenButton from './FullScreenButton.component';

describe('FullScreenButton tests', () => {
  test('default renderer', () => {
    shallowMatchSnapshot(<FullScreenButton />);
  });
});
