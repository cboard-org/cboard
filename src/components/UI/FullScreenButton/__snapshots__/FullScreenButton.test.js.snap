// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FullScreenButton tests default renderer 1`] = `
<FullScreenButton
  intl={
    Object {
      "formatMessage": [Function],
    }
  }
/>
`;
